### ISSUE TITLE
Problems binding data at runtime

### ISSUE THREAD

**Issue by jefffriesen** [uwdata/vega#57]
This may have broken in a recent update (or I might be doing something wrong of course): https://github.com/jefffriesen/vega/blame/webpack/src/core/View.js#L92

When I try to pass in data into the config object, I get errors. Here's the problem code with the screenshot and error message. 
#### Problem

```
let data = [
  {"x": 1,  "y": 28}, {"x": 2,  "y": 55},
  {"x": 3,  "y": 43}, {"x": 4,  "y": 91},
  {"x": 5,  "y": 81}, {"x": 6,  "y": 53},
  {"x": 7,  "y": 19}, {"x": 8,  "y": 87},
  {"x": 9,  "y": 52}, {"x": 10, "y": 48},
  {"x": 11, "y": 24}, {"x": 12, "y": 49},
  {"x": 13, "y": 87}, {"x": 14, "y": 66},
  {"x": 15, "y": 17}, {"x": 16, "y": 27},
  {"x": 17, "y": 68}, {"x": 18, "y": 16},
  {"x": 19, "y": 49}, {"x": 20, "y": 15}
]
...
let barChartConfig = {
  "width": 400,
  "height": 200,
  "padding": {"top": 10, "left": 30, "bottom": 30, "right": 10},
  "data": [
    {
      "name": "table"
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "ordinal",
      "range": "width",
      "domain": {"data": "table", "field": "x"}
    },
    {
      "name": "y",
      "range": "height",
      "nice": true,
      "domain": {"data": "table", "field": "y"}
    }
  ],
  "axes": [
    {"type": "x", "scale": "x"},
    {"type": "y", "scale": "y"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "table"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "x"},
          "width": {"scale": "x", "band": true, "offset": -1},
          "y": {"scale": "y", "field": "y"},
          "y2": {"scale": "y", "value": 0}
        },
        "update": {
          "fill": {"value": "steelblue"}
        },
        "hover": {
          "fill": {"value": "red"}
        }
      }
    }
  ]
}
...
vega.parse.spec(barChartConfig, function(chart) {
    chart({el:node})
        .data({"table": data})
        .update()

})
```

![screen shot 2015-05-24 at 12 06 18 am](https://cloud.githubusercontent.com/assets/445733/7786926/1c389132-01aa-11e5-92de-1b5959e6d6c0.png)

The error is

```
Uncaught TypeError: data[k] is not a function
```
#### Working:

If I pass in data to the config object directly it works fine:

```
let data = {
   {"x": 1,  "y": 28}, {"x": 2,  "y": 55},
  ...
}
...
"data": [
    {
      "name": "table",
      "values": data
    }
],
...
```

I wanted to update the charts based on user interaction, so I need to bind and update the charts during runtime.

Thanks

**Comment by arvind** [comment]
Hi Jeff, 

Apologies for the slow response, I wanted to get some documentation up before replying. This is an intended breaking change to the API in order to support streaming data. Until I write documentation to better describe the new API, you should hopefully be able to piece together what you need using [this guide](https://github.com/uwdata/vega/wiki/Upgrading-to-2.0#runtime-view-api) and PR #48. Let me know if you have any questions!

**Comment by jefffriesen** [comment]
Thanks for the update to the docs. I still haven't been able to make it work yet. 

In the new documentation (https://github.com/uwdata/vega/wiki/Upgrading-to-2.0#runtime-view-api), it's not clear to me what `view` is. I think it's the `chart` reference shown in the original runtime documentation: https://github.com/uwdata/vega/wiki/Runtime#basic-use

```
vg.parse.spec(spec, function(chart) { 
  chart({el:"#vis"}).update(); 
});
```

Here is what I tried based your docs: (not working)

```
vega.parse.spec(barChartConfig, function(chart) {
  chart({el:node})
    .data("table")
    .remove(function(d) { return true; }) 
    .insert(values)
    // I tried to chain an .update() and a .fire() here with no luck
})
```

There's no error, just no chart. I looked in the tests to see if there were examples but didn't find any examples that showed me my problem. I think my setup is otherwise correct, because when I set the `values` array directly in the `barChartConfig` object, it works fine using the non-runtime binding:

```
vega.parse.spec(barChartConfig, function(chart) {
  chart({el:node}).update()
})
```

(btw, I'm excited you are taking an RxJS-type approach to interactions. It's one of the reasons that drew me to exploring this library more)

**Comment by meetamit** [comment]
I believe this is the proper way:

```
vega.parse.spec(barChartConfig, function(chart) {
  var view = chart({el:node});
  view
    .data("table")
    .remove(function(d) { return true; }) 
    .insert(values);
  view.update();
})
```

That's because it is necessary to update the view instance after inserting the data. Note that the `insert(values)` call does not return the instance of `view`, which I think is why your attempt to chain `.update()` didn't work.

**Comment by jefffriesen** [comment]
Thanks @meetamit and @arvind 
Looking at your examples and source code I have a better understanding of things, but I still haven't been able to make it work.

I tried 

```
vega.parse.spec(barChartConfig, function(chart) {
  var view = chart({el:node});
  view.data("table")
    .remove(function(d) { return true; }) 
    .insert(values);
  view.update()
})
```

and

```
vega.parse.spec(barChartConfig, function(chart) {
  var view = chart({el:node});
  view.data({
      table: function(data) {
        data.remove(function(d) { return true; })
          .insert(values);
      }
    })
  view.update()
})
```

I tried `values` as a JS array and as an RxJS Observable (`Rx.Observable.from(values)`) , neither would work (no error, but just got axis and no chart data). Again, I think the rest of my setup is ok, because I can get the chart rendering if I set the `values` on the `barChartConfig` object. 

I think one or two full examples would clear up a lot, but I know a lot is in flux right now and you're busy still putting it together. Thanks for the work you've all done on this.

**Comment by arvind** [comment]
You're right, it seems like a regression was introduced at some point.

**Comment by arvind** [comment]
@jefffriesen thanks for the bug report, this should be fixed in 06d8d43. Let me know if you still run into any issues. 

``` js
let data = [
  {"x": 1,  "y": 28}, {"x": 2,  "y": 55},
  {"x": 3,  "y": 43}, {"x": 4,  "y": 91},
  {"x": 5,  "y": 81}, {"x": 6,  "y": 53},
  {"x": 7,  "y": 19}, {"x": 8,  "y": 87},
  {"x": 9,  "y": 52}, {"x": 10, "y": 48},
  {"x": 11, "y": 24}, {"x": 12, "y": 49},
  {"x": 13, "y": 87}, {"x": 14, "y": 66},
  {"x": 15, "y": 17}, {"x": 16, "y": 27},
  {"x": 17, "y": 68}, {"x": 18, "y": 16},
  {"x": 19, "y": 49}, {"x": 20, "y": 15}
]
...
let barChartConfig = {
  "width": 400,
  "height": 200,
  "padding": {"top": 10, "left": 30, "bottom": 30, "right": 10},
  "data": [
    {
      "name": "table"
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "ordinal",
      "range": "width",
      "domain": {"data": "table", "field": "x"}
    },
    {
      "name": "y",
      "range": "height",
      "nice": true,
      "domain": {"data": "table", "field": "y"}
    }
  ],
  "axes": [
    {"type": "x", "scale": "x"},
    {"type": "y", "scale": "y"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "table"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "x"},
          "width": {"scale": "x", "band": true, "offset": -1},
          "y": {"scale": "y", "field": "y"},
          "y2": {"scale": "y", "value": 0}
        },
        "update": {
          "fill": {"value": "steelblue"}
        },
        "hover": {
          "fill": {"value": "red"}
        }
      }
    }
  ]
}

vega.parse.spec(barChartConfig, function(chart) {
  var view = chart({el:node});
  view.data("table")
    .insert(values);
  view.update()
})
```
